ocs_ci/ocs/version.py

```diff
@@ -0,0 +1,174 @@
+   given pod is running)
+"""
+
+
+import argparse
+import logging
+import os.path
+import pprint
+import re
+import sys
+
+from ocs_ci import framework
+from ocs_ci.framework import config
+from ocs_ci.ocs.ocp import OCP
+from ocs_ci.utility import utils
+
+
+logger = logging.getLogger(__name__)
+
+
+def get_ocs_version():
+    """
+    Query OCP to get all information about OCS version.
+
+    Returns:
+     * dict with ClusterVersion k8s object
+     * image_dict with information about images IDs
+    """
+    logger.info("collecting ocs version")
+
+    # We use ClusterVersion resource (which is acted upon by openshift
+    # cluster-version operator, aka CVO) to get the version of our OCP
+    # instance. See also:
+    # https://github.com/openshift/cluster-version-operator/blob/master/docs/dev/clusterversion.md
+    ocp = OCP(kind="clusterversion")
+    cluster_version = ocp.get("version")
+
+    # TODO: When OLM (operator-lifecycle-manager) maintains OCS, it will be
+    # possible to add a check via CSV (cluster service version) to get OCS
+    # version in a similar way as it's done for OCP itself above.
+    # Reference: Jose A. Rivera on ocs-qe list.
+
+    # Get all openshift storage related namespaces. Eg. at the time of writing
+    # this code (July 2019), there were these storage namespaces:
+    #  * openshift-cluster-storage-operator
+    #  * openshift-storage
+    # TODO: how to do this in upstream where namespace is rook-ceph?
+    # TODO: should we use defaults.ROOK_CLUSTER_NAMESPACE somehow here?
+    storage_namespaces = []
+    all_namespaces = OCP(kind="namespace").get()['items']
+    ns_re = re.compile("^openshift.*storage")
+    for ns in all_namespaces:
+        if (ns_re.match(ns['metadata']['name'])):
+            storage_namespaces.append(ns['metadata']['name'])
+
+    logger.info("found storage namespaces %s", storage_namespaces)
+
+    # Now get the OCS version by asking for version of all container images of
+    # all pods in openshift-storage namespace.
+    image_dict = {}
+    for ns in storage_namespaces:
+        ocs = OCP(kind="pod", namespace=ns)
+        ocs_pods = ocs.get()
+        ns_dict = {}
+        for pod in ocs_pods['items']:
+            for container in pod['spec']['containers']:
+                logger.debug(
+                    'container %s (in pod %s) uses image %s',
+                    container['name'],
+                    pod['metadata']['name'],
+                    container['image'])
+            for cs in pod['status']['containerStatuses']:
+                ns_dict.setdefault(cs['image'], set()).add(cs['imageID'])
+        image_dict.setdefault(ns, ns_dict)
+
+    logger.debug("ocs version collected")
+
+    return cluster_version, image_dict
+
+
+def report_ocs_version(cluster_version, image_dict, file_obj):
+    """
+    Report OCS version via:
+
+     * python logging
+     * printing human readable version into file_obj (stdout by default)
+    """
+    # log the version
```

Comment:
Sorry missed you question about report in some machine readable format.

I think that if we will write the data in the dictionary like:

report_data= {}
report_data[cluster_channel] = cluster_version["spec"]["channel"]
....

that we can at the end dump the yaml file which is human readable just fine and can be printed out.
This is not need for this PR now , just thinking about usage, currently the email reporting is done in:
`ocs_ci/framework/pytest_customization/reports.py`
and if we would like to attach this report to the email we will have to do there and read those data generated here so we will need to make sure that this version report is done sooner than email reporting to be able to read the data from somewhere (file or return value or so..)